plugins/state_history_plugin/session.hpp

```diff
@@ -0,0 +1,421 @@
+namespace ws  = boost::beast::websocket;
+namespace bio = boost::iostreams;
+
+template <typename Session>
+struct send_queue_entry_base {
+   virtual ~send_queue_entry_base() {}
+   virtual void send(Session* s) = 0;
+};
+
+template <typename Session>
+struct basic_send_queue_entry : send_queue_entry_base<Session> {
+   std::vector<char> data;
+   template <typename... Args>
+   basic_send_queue_entry(Args&&... args)
+       : data(std::forward<Args>(args)...) {}
+   void send(Session* s) override {
+      s->socket_stream.async_write(boost::asio::buffer(data),
+                                   [s = s->shared_from_this()](boost::system::error_code ec, size_t) {
+                                      s->callback(ec, "async_write", [s] { s->pop_entry(); });
+                                   });
+   }
+};
+
+template <typename Session>
+class blocks_result_send_queue_entry : public send_queue_entry_base<Session> {
+   eosio::state_history::get_blocks_result_v0                    r;
+   std::vector<char>                                             data;
+   std::variant<std::vector<char>, maybe_locked_compress_stream> buf;
+
+   template <typename Next>
+   void async_send(Session* s, bool fin, const std::vector<char>& data, Next&& next) {
+      s->socket_stream.async_write_some(
+          fin, boost::asio::buffer(data),
+          [s = s->shared_from_this(), next = std::forward<Next>(next)](boost::system::error_code ec, size_t) mutable {
+             s->callback(ec, "async_write", [s, next = std::move(next)]() mutable { next(s.get()); });
+          });
+   }
+
+   template <typename Next>
+   void async_send(Session* s, bool fin, maybe_locked_compress_stream& locked_strm, Next&& next) {
+      data.resize(s->default_frame_size);
+      auto size = bio::read(locked_strm.buf, data.data(), s->default_frame_size);
+      data.resize(size);
+      bool eof = (locked_strm.buf.sgetc() == EOF);
+
+      s->socket_stream.async_write_some(
+          fin && eof, boost::asio::buffer(data),
+          [this, s = s->shared_from_this(), fin, &locked_strm, eof,
+           next = std::forward<Next>(next)](boost::system::error_code ec, size_t) mutable {
+             if (ec) locked_strm.lock.reset();
+             s->callback(ec, "async_write", [this, s, fin, &locked_strm, eof, next = std::move(next)]() mutable {
+                if (eof) {
+                   locked_strm.lock.reset();
+                   next(s.get());
+                } else {
+                   async_send(s.get(), fin, locked_strm, std::move(next));
+                }
+             });
+          });
+   }
+
+   template <typename Next>
+   void async_send(Session* s, bool fin, std::variant<std::vector<char>, maybe_locked_compress_stream>& buf,
+                  Next&& next) {
+      std::visit([this, s, fin,
+                  next = std::forward<Next>(next)](auto& data) mutable { this->async_send(s, fin, data, std::move(next)); },
```

Comment:
The param name of `data` is very confusing with the member data. Maybe use `d` here instead.